packages/controllers/src/multichain/MultiChainController.ts

```diff
@@ -0,0 +1,453 @@
+  handlingSnaps: Record<NamespaceId, SnapId>;
+};
+
+type MultiChainControllerState = {
+  sessions: { [origin: string]: SessionData };
+};
+
+type Notify = (
+  origin: string,
+  data: { method: string; params?: Record<string, unknown> },
+) => Promise<void>;
+
+type MultiChainControllerArgs = {
+  notify: Notify;
+  messenger: MultiChainControllerMessenger;
+};
+
+/**
+ * Parse a chain ID string to an object containing the namespace and reference.
+ * This validates the chain ID before parsing it.
+ *
+ * @param chainId - The chain ID to validate and parse.
+ * @returns The parsed chain ID.
+ */
+export function parseChainId(chainId: ChainId): {
+  namespace: NamespaceId;
+  reference: string;
+} {
+  if (!isChainId(chainId)) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  const match = CHAIN_ID_REGEX.exec(chainId);
+  if (!match?.groups) {
+    throw new Error('Invalid chain ID.');
+  }
+
+  return {
+    namespace: match.groups.namespace,
+    reference: match.groups.reference,
+  };
+}
+
+/**
+ * Parse an account ID to an object containing the chain, chain ID and address.
+ * This validates the account ID before parsing it.
+ *
+ * @param accountId - The account ID to validate and parse.
+ * @returns The parsed account ID.
+ */
+export function parseAccountId(accountId: AccountId): {
+  chain: { namespace: NamespaceId; reference: string };
+  chainId: ChainId;
+  address: string;
+} {
+  if (!isAccountId(accountId)) {
+    throw new Error('Invalid account ID.');
+  }
+
+  const match = ACCOUNT_ID_REGEX.exec(accountId);
+  if (!match?.groups) {
+    throw new Error('Invalid account ID.');
+  }
+
+  return {
+    address: match.groups.accountAddress,
+    chainId: match.groups.chainId as ChainId,
+    chain: {
+      namespace: match.groups.namespace,
+      reference: match.groups.reference,
+    },
+  };
+}
+
+// TODO(ritave): Support for legacy ethereum operations, not just snaps
+export class MultiChainController extends BaseController<
+  typeof controllerName,
+  MultiChainControllerState,
+  MultiChainControllerMessenger
+> {
+  private notify: Notify;
+
+  constructor({ messenger, notify }: MultiChainControllerArgs) {
+    super({
+      messenger,
+      metadata: {
+        sessions: { persist: false, anonymous: false },
+      },
+      name: controllerName,
+      state: defaultState,
+    });
+
+    this.notify = notify;
+  }
+
+  getSession(origin: string): SessionData | undefined {
+    return this.state.sessions[origin];
+  }
+
+  async closeSession(origin: string): Promise<void> {
+    const session = this.getSession(origin);
+    assert(session, new Error('No session to close'));
+
+    await this.notify(origin, { method: 'multichainHack_metamask_disconnect' });
+
+    this.update((state) => {
+      delete state.sessions[origin];
+    });
+
+    await Promise.all(
+      Object.values(session.handlingSnaps).map((snapId) =>
+        this.messagingSystem.call('SnapController:onSessionClose', snapId),
```

Comment:
I'd prefer if the SnapController didn't know about the concept of a Session at all

Reply:
I imagined SnapController "sessions" being a multi-purpose abstraction that we _could_ use in the future if we wanted a snap used by a DApp to be kept alive for the entire "DApp session", not just for keyrings.

Reply:
I like the idea. Depends if we want to have `SnapController` be smart about it, or just be a dump reference counting controller.

SnapController is big enough, and I believe it should only manage parts that concerning the actual execution of snaps, not higher level logic of sessions. Personally I'd like it to just do reference counting (`SnapController:incOpenReferences` / `SnapController:decOpenReferences`) and close on zero references. But I'm open to discussion here

Reply:
It is basically a dumb reference counter at this point. It is merely the question of the name. I like "open sessions" better, but could be persuaded to another name 😅

Reply:
I'd like anything **but** sessions to avoid name clashing